### PR TITLE
Allow overriding cache directory through config

### PR DESCRIPTION
**Pull request by dbrgn**
Based on #212. Thanks @hgaiser for the initial work!

- Config: Add `cache_dir` to `[directories]` section
- Deprecate old `TEALDEER_CACHE_DIR` env variable
- Introduce path source for `custom_pages_dir` as well
- Add integration test
- Update docs

Implementation notes:

- Logic to handle legacy env var, config override and default cache dir selection is now in the `Config::try_from(RawConfig)` impl
- Logic to determine `custom_pages_dir` is moved into `Config::try_from(RawConfig)` as well for consistency

The logic to determine which cache directory is used is now in `config.rs`, when converting the `RawConfig` to `Config`. I think this is the place where the config file and external factors (defaults, env vars, etc) can be taken into account to determine a final config.

### CODE REVIEW

**Comment by dbrgn** [comment]
@niklasmohrin sorry for missing #266, I wanted to get PRs unblocked starting at the oldest, and didn't yet see that you already submitted a proposed solution (I skimmed over it since it was still marked as draft)... 🙁

I think I chose a different approach than you did. I determine the final path in `Config::try_from(RawConfig)`. While implementing, I felt that this seems to be a clean solution, because once you get a valid config, you have a final non-nullable config path.

Validation of that path is being done by the cache itself: If the cache is instantiated with a path, it tries to ensure that the path exists. If it doesn't, it is created. If creation fails, the program is aborted.

I'd be interested in your opinion on that.

File: src/config.rs [hunk 1]
```diff
@@ -249,30 +252,93 @@ pub struct Config {
...
+    /// defaults need to be set.
+    fn try_from(raw_config: RawConfig) -> Result<Self, Self::Error> {
+        // Style config
+        let style = StyleConfig {
+            command_name: raw_config.style.command_name.into(),
+            description: raw_config.style.description.into(),
+            example_text: raw_config.style.example_text.into(),
+            example_code: raw_config.style.example_code.into(),
+            example_variable: raw_config.style.example_variable.into(),
+        };
```
Comment (niklasmohrin) [inline src/config.rs:270]: Should we have `StyleConfig: From<RawStyleConfig`?

File: docs/src/config_directories.md [hunk 2]
```diff
@@ -2,6 +2,19 @@
 
 This section allows overriding some directory paths.
 
+## `cache_dir`
+
+Override the cache directory. Remember to use an absolute path. Variable
+expansion will not be performed on the path. If the directory does not yet
+exist, it will be created.
```
[the remark this answers is not in the document]
Comment (dbrgn) [reply docs/src/config_directories.md:9]: Don't you think it's still worthwhile to document that this is the expected behavior? For me, it would indicate that this behavior is by design, not by concidence. (If you still think the sentence should be removed, I can do that.)

File: src/cache.rs [hunk 3]
```diff
@@ -89,49 +87,42 @@ pub enum CacheFreshness {
...
+            // Try to create the complete directory path
+            fs::create_dir_all(&cache_dir).with_context(|| {
+                format!(
+                    "Cache directory path `{}` cannot be created",
+                    cache_dir.display()
+                )
+            })?;
+            eprintln!(
+                "Successfully created cache directory path `{}`.",
+                cache_dir.display()
```
[the remark this answers is not in the document]
Comment (dbrgn) [reply src/cache.rs:117]: Yeah, rustfmt doesn't add automatically them becuase (e)println is a macro.

File: src/cache.rs [hunk 4]
```diff
@@ -25,6 +22,7 @@ static TLDR_OLD_PAGES_DIR: &str = "tldr-master";
 pub struct Cache {
     url: String,
```
[the remark this answers is not in the document]
Comment (dbrgn) [reply src/cache.rs]: I don't have a strong opinion, but your suggestion probably makes sense. I moved the URL to a parameter.

File: src/cache.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbrgn) [reply src/cache.rs]: Should all be fixed now.

File: src/cache.rs [hunk 5]
```diff
@@ -302,10 +280,9 @@ impl Cache {
     }
 
     /// Return the available pages.
-    pub fn list_pages(&self) -> Result<Vec<String>> {
+    pub fn list_pages(&self) -> Vec<String> {
         // Determine platforms directory and platform
-        let (cache_dir, _) = Self::get_cache_dir()?;
-        let platforms_dir = cache_dir.join(TLDR_PAGES_DIR).join("pages");
+        let platforms_dir = self.cache_dir.join(TLDR_PAGES_DIR).join("pages");
```
[the remark this answers is not in the document]
Comment (dbrgn) [reply src/cache.rs]: Good call, done!

File: src/config.rs (hunk 1 above)
Comment (dbrgn) [reply src/config.rs]: Doesn't hurt! Added.

**Comment by dbrgn** [comment]
@niklasmohrin everything should be updated and all your feedback should be addressed!

File: docs/src/config_directories.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (dbrgn) [reply docs/src/config_directories.md:9]: Actually, I'd keep the sentence. Intuitively, as a user, I would expect the cache directory to be created when downloading the cache, or that an error message would be returned if it doesn't yet exist. However, it's created as soon as tealdeer is run:

screenshot-20220924-221721

[the remark this answers is not in the document]
Comment (niklasmohrin) [reply docs/src/config_directories.md:9]: Hm, I don't know what motivated my initial comment, probably felt like it was more "feature-bragging" than useful, but you are right that it makes sense to have documentation to point to if questions arise.

However, I think that your screenshot is a misbehavior, we shouldn't create it unless we need to, or? Let's fix that before the next release (if not in this PR :D)

**Review by niklasmohrin — CHANGES_REQUESTED** [review]
Nice! Some more comments, a handful of them are suggestions about removing code comments - I think that the code following those comments is just as expressive (if not more) than the comment while being more precise. I think the comments here add noise and don't provide any benefit. (and, also, all the textbook arguments like "comments aren't run in tests", "comments duplicate knowledge", ...)

There are more such comments around the codebase (and I would, in most cases, like them gone; not today, but with every little refactor :D). If I recall correctly, we disagreed over this before, haven't we? What did we settle on back then?

**Comment by dbrgn** [comment]
> Nice! Some more comments, a handful of them are suggestions about removing code comments - I think that the code following those comments is just as expressive (if not more) than the comment while being more precise. I think the comments here add noise and don't provide any benefit. (and, also, all the textbook arguments like "comments aren't run in tests", "comments duplicate knowledge", ...)
> 
> There are more such comments around the codebase (and I would, in most cases, like them gone; not today, but with every little refactor :D). If I recall correctly, we disagreed over this before, haven't we? What did we settle on back then?

Hehe, I know that discussion from other projects 🙂 It comes down to a matter of style, but I think there are good reasons for it. I really like to structure code into blocks that do some kind of action. For example, when creating a file in a directory:

```rust
fn myfunction() {
    check_if_dir_exists();
    check_if_dir_is_writable();
    write_file();
    set_file_permissions();
}
```

In the example above, the three function calls are only 1 line each and self-explanatory. However, when code get bigger, sometimes you need to fully read through the block to understand what it does. With trivial code, that's simple. But if it's a bit more complex, it takes longer.

So I follow a code style where these blocks (potentially spanning multiple lines of code) are preceded with a line comment that explains what the following code does. For example:

```rust
fn myfunction() {
    // Ensure that we can write the directory
    check_if_dir_exists();
    check_if_dir_is_writable();

    // Directory is writable! Actually write the file
    write_file();
    set_file_permissions();
}
```

What I like about this style is that it's really simple to skim over an implementation by just looking at the block headers. Sometimes, blocks are small and self-explanatory, but I usually add these comments anyways for consistency. The idea is not to just state what the next line is doing, but to guide the reader through the code, and - if it makes sense - to also explain the reason why certain things are the way they are.

As a concrete example. This code:

screenshot-20220928-000532

...can be skimmed over easily, especially with syntax highlighting. To my mind, I read it like this:

screenshot-20220928-000649

...which, I would argue, is much easier to quickly understand than this:

screenshot-20220928-000729

Note that the last example is _not_ unreadable. I think it's fairly easy to read. But it takes more time and brainpower to process than the version where each block is summarized.

I do kind of feel strongly about this, so I would not want to take away comments, as long as this "block-summary-comment" style is consistent within a function. I don't expect contributors to do this as well, of course, as long as functions remain small and easy to read. (Once functions do get more complex, I think it's valuable to either split them up into multiple functions, or to use comments in a more structured way.)

File: src/config.rs [hunk 6]
```diff
@@ -249,30 +273,84 @@ pub struct Config {
...
+        // Display config
+        let display = raw_config.display.into();
+
+        // Updates config
+        let updates = UpdatesConfig {
+            auto_update: raw_config.updates.auto_update,
+            auto_update_interval: Duration::from_secs(
+                raw_config.updates.auto_update_interval_hours * 3600,
+            ),
+        };
```
[the remark this answers is not in the document]
Comment (dbrgn) [reply src/config.rs:296]: Ah, I looked at it and thought "we need to do some custom processing on that, with the `auto_update_interval`, but in the end it's just a type conversion. I'll add a `From` impl for this one as well!

File: src/config.rs [hunk 7]
```diff
@@ -249,30 +273,84 @@ pub struct Config {
...
+    /// Convert a `RawConfig` to a high-level `Config`.
+    ///
+    /// For this, some values need to be converted to other types and some
+    /// defaults need to be set.
+    fn try_from(raw_config: RawConfig) -> Result<Self, Self::Error> {
+        // Style config
+        let style = raw_config.style.into();
+
+        // Display config
+        let display = raw_config.display.into();
```
[the remark this answers is not in the document]
Comment (dbrgn) [reply src/config.rs:288]: Good call, here it absolutely makes sense to collapse the blocks, and it looks a bit silly otherwise 😄 It's a leftover from the refactoring / simplification.

File: src/cache.rs [hunk 8]
```diff
@@ -258,15 +241,8 @@ impl Cache {
...
-        let cache_dir = match Self::get_cache_dir() {
-            Ok((cache_dir, _)) => cache_dir.join(TLDR_PAGES_DIR),
-            Err(e) => {
-                log::error!("Could not get cache directory: {}", e);
-                return None;
-            }
-        };
-
+        // Determine directories
+        let pages_dir = self.pages_dir();
```
[the remark this answers is not in the document]
Comment (dbrgn) [reply src/cache.rs:245]: Yep, I fixed the naming. I'll check (and fix) the argument as well.

**Comment by niklasmohrin** [comment]
If you say you have a strong opinion about this, I can live with it, although my opinion isn't changed. Here is how I see it:

---

Firstly, I can see that this is useful for reading complex functions, but to me it's unclear where to draw the line. You agreed at the "style config" comment, but I think that
```rust
        // Extract archive into pages dir
        archive
            .extract(&self.pages_dir())
            .context("Could not unpack compressed data")?;
```

and this snippet from main:

```rust
// Initialize logger
init_log();

// Parse arguments
let mut args = Args::parse();
```
look "just as bad" to me. (tbh, I think that all of `Cache::update` is fine without these heading comments).

I find that having a block summarized as just exactly what is already said in the block is just more noisy, so I would first change to something like this:
grafik

Now, I too find this a bit confusing. For example, after the renamed args, we have another `if args.something` but that does logic rather than cleanup and one could think that it still belongs to the block before because there is no new block heading (so the _absence_ of a heading confused me). While adding the heading back would help with that slip-up, I think it is not the correct approach. What really happened is that our `main` function just doesn't stick to one level of abstraction and we are trying to patch this up by bundling code into blocks which are then on the same abstraction level. For simple code that is correctly abstracted, this leads to `/* init log */ init_log();`. So what we should aim for is:
grafik

I guess what I am trying to say is: I can understand the "summarizing the block" part, but I don't fully agree with the "make usage of block comments consistent within function". To me, it is a symptom of violating the single-level-of-abstraction rule. And yes, it is somewhat eased with helping comments, but in reality we should just try harder to find better words for what the code should do. And maybe, these comments are just a marker for a refactor later, because we are trying to be productive now. But that's it for me.

---

Again, at the end of the day we want to get a feature shipped here and I wouldn't want to have this blocked over a fight about comments (they literally do nothing :D). I am curious how you feel about this though

File: docs/src/config_directories.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (dbrgn) [reply docs/src/config_directories.md:9]: The advantage of the current solution is that we never have to check for existence of the cache directory before we use it, we initialize it in the beginning and then assume it's there. This simlifies some code paths.

However, after second thought I agree that this might not be ideal from a user perspective. Additionally, we always need to expect races on the file system. I changed it to create the cache directory lazily.

File: src/config.rs [hunk 9]
```diff
@@ -249,30 +273,84 @@ pub struct Config {
...
-                auto_update: raw_config.updates.auto_update,
-                auto_update_interval: Duration::from_secs(
-                    raw_config.updates.auto_update_interval_hours * 3600,
-                ),
-            },
-            directories: DirectoriesConfig {
-                custom_pages_dir: raw_config.directories.custom_pages_dir,
-            },
-        }
+impl TryFrom<RawConfig> for Config {
```
[the remark this answers is not in the document]
Comment (dbrgn) [reply src/config.rs:277]: Fair point. I renamed it to `from_raw` and expanded the function docs.

File: src/config.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (dbrgn) [reply src/config.rs:297]: Added a `From` for this type as well.

The code could be refactored further (like moving the updates config conversion to a dedicated function), but I'll stop with refactoring for now.

File: tests/lib.rs [hunk 10]
```diff
@@ -263,9 +272,53 @@ fn test_cache_location_not_a_directory() {
...
+    command.env(CACHE_DIR_ENV_VAR, default_cache_dir.to_str().unwrap());
+    command
+        .arg("--show-paths")
+        .assert()
+        .success()
+        .stdout(is_match("\nCache dir:        [^(]* \\(env variable\\)\n").unwrap());
+
+    // Source: Config variable
+    let mut command = testenv.command();
+    command.env_remove(CACHE_DIR_ENV_VAR);
```
[the remark this answers is not in the document]
Comment (dbrgn) [reply tests/lib.rs:312]: Good point!

**Comment by dbrgn** [comment]
Great, thanks for the review. All issues should have been addressed.

Regarding the comments: Yes, absolutely, often the clarity of the code can be improved with refactoring or better naming. I'm not saying that shouldn't be done, on the contrary 🙂 And if a method is simple enough, then comments aren't needed. Once it crosses a certain complexity, I simply like adding block comments (in the style of headlines) consistently. Even if this means that sometimes very short blocks have a comment as well. But it looks more consistent.

Any other last feedback? I plan to merge this and prepare the 1.5.1 release by the end of the week.

File: src/config.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (niklasmohrin) [reply src/config.rs]: @dbrgn I am curious: Do you mean adding a `UpdatesConfig::from_raw_config`? I think having the logic in the `From` is the most idiomatic approach, not needing any more indirection

File: src/cache.rs [hunk 11]
```diff
@@ -89,53 +86,57 @@ pub enum CacheFreshness {
...
-            let (path_exists, path_is_dir) = path
-                .metadata()
-                .map_or((false, false), |md| (true, md.is_dir()));
-            ensure!(
-                !path_exists || path_is_dir,
-                "Path specified by ${} is not a directory",
-                CACHE_DIR_ENV_VAR
+    pub fn cache_dir(&self) -> &Path {
+        &self.cache_dir
+    }
```
Comment (niklasmohrin) [inline src/cache.rs:101]: Was this here before? Since we are always using `self.cache_dir` as the implicit `&Path` anyways, let's replace all usages of `cache_dir` with `cache_dir()`

File: src/cache.rs [hunk 12]
```diff
@@ -367,29 +353,24 @@ impl Cache {
...
-
-        // Check preconditions
-        ensure!(
-            path.exists(),
-            "Cache path ({}) does not exist.",
-            path.display(),
-        );
+    pub fn clear(&self) -> Result<bool> {
+        if !self.cache_dir.exists() {
+            return Ok(false);
```
Comment (niklasmohrin) [inline src/cache.rs:362]: We should add an explanation of the `bool` in the doc comment

File: src/config.rs [hunk 13]
```diff
@@ -1,11 +1,12 @@
 use std::{
-    env, fs,
+    convert::TryFrom,
```
Comment (niklasmohrin) [inline src/config.rs:2]: This is unused

File: src/main.rs [hunk 14]
```diff
@@ -393,7 +380,11 @@ fn main() {
         if let Some(lookup_result) = cache.find_page(
             &command,
             &languages,
-            config.directories.custom_pages_dir.as_deref(),
+            config
+                .directories
+                .custom_pages_dir
+                .as_ref()
+                .map(|path_with_source| path_with_source.path.as_ref()),
```
Comment (niklasmohrin) [inline src/main.rs:387]: This looks a bit odd, it can be more concise if we implement a `&Path` getter on `PathWithSource`:
```suggestion
            config
                .directories
                .custom_pages_dir
                .as_ref()
                .map(PathWithSource::path),
```
(same 20 lines above)

**Review by niklasmohrin — APPROVED** [review]
Nice! Some last remarks for this PR:

File: src/cache.rs (hunk 11 above)
Comment (dbrgn) [reply src/cache.rs:101]: You can see the result of the force push here 🙂 https://github.com/dbrgn/tealdeer/compare/8c8f612e38e247843c4bb243520645e4147ed4eb..2894f4823c3a55a02d5b34f4d50d6077ab1038af (On GitHub, click the "force-pushed" text in the message "dbrgn force-pushed the cache-dir-config branch from 8c8f612 to 2894f48 2 days ago" to see the diff.)

The method was here before. But it is only needed to expose the cache directory towards the outside (because the `cache_dir` field is not accessible). I disagree that we should go through a method everywhere just to access a field with the same name, both from an abstraction level point of view and based on the interest of getting this merged without too much additional bikeshedding.

Comment (niklasmohrin) [reply src/cache.rs:101]: Fair enough 🤷

**Comment by dbrgn** [comment]
🎉 Thanks! All feedback should now be addressed.
